### PR TITLE
[sqllogictest] Define output types and check them in tests

### LINKED ISSUE #4499: Implement column number / column type verification for sqllogictest
https://duckdb.org/dev/sqllogictest/result_verification

> The standard way of verifying results of queries is using the query statement, followed by the letter I times the number of columns that are expected in the result.

```sql
query II
SELECT 42, 84 UNION ALL SELECT 10, 20;
----
42	84
10	20
```

At the moment, the datafusion sqllogictest framework entirely ignores the number of `I` after the `query` command

The goal of this ticket is to update the framework (and tests) to error if the number of `I`s doesn't match the number of columns

### PR DESCRIPTION
**Pull request by melgenek**
# Which issue does this PR close?

Closes #4499.

# Rationale for this change

Currently, the type strings in sqllogictests are ignored, and letters have no meaning to them. Also, some tests benefit, like null checks, benefit from type checking.

# What changes are included in this PR?

- this pr defines a custom `DFColumnType` that has more types than the original sqllite set of types
- updates tests to use these custom types
- forces the sqllogictestrunner to check the types

# Are these changes tested?

Github actions.

# Are there any user-facing changes?

No

### CODE REVIEW

File: datafusion/core/tests/sqllogictests/README.md [hunk 1]
```diff
@@ -105,10 +105,13 @@ query <type_string> <sort_mode>
...
-  - "I" for an integer result,
-  - "R" for a floating-point result,
-  - "?" for any other type.
+  - 'B' - **B**oolean,
+  - 'D' - **D**atetime,
+  - 'I' - **I**nteger,
+  - 'F' - **F**loating-point results,
+  - 'S' - **S**string,
+  - 'T' - **T**imestamp,
+  - "?" - any other types
```
Comment (melgenek) [inline datafusion/core/tests/sqllogictests/README.md:114]: Honestly, I wanted to add only `Boolean` at first but then reshuffled the letters to accommodate more types.

Notable differences here:
- `F` for the floating points. Sqllite defines floating points as `R`. [Cocroach](https://github.com/cockroachdb/cockroach/blob/899c9112188c22ec140adbeb0cf7d5761da3827d/pkg/sql/logictest/logic.go#L285-L287) differentiates floating points `F`, and decimals `R`
- `S` for strings. Sqllite and Cocroach define these as `T` text. But I repurposed `T` for **t**imestamps.

I tried using first letters of the words that have some meaning, but I am open for suggestions here.

File: datafusion/core/tests/sqllogictests/src/engines/datafusion/mod.rs [hunk 2]
```diff
@@ -18,7 +18,7 @@
 use std::path::PathBuf;
 use std::time::Duration;
 
-use crate::output::{DFColumnType, DFOutput};
+use crate::engines::output::{DFColumnType, DFOutput};
```
Comment (melgenek) [inline datafusion/core/tests/sqllogictests/src/engines/datafusion/mod.rs:21]: I moved the `output.rs` into the `engines/` folder. I just prefer narrowing the scope of files to where they are used, and there's no need to expose the type outside of the `engines/` folder now.

File: datafusion/core/tests/sqllogictests/src/engines/postgres/mod.rs [hunk 3]
```diff
@@ -319,3 +271,73 @@ impl sqllogictest::AsyncDB for Postgres {
...
+    ($row:ident, $idx:ident, $t:ty, $convert:ident) => {{
+        let value: Option<$t> = $row.get($idx);
+        match value {
+            Some(value) => $convert(value).to_string(),
+            None => NULL_STR.to_string(),
+        }
+    }};
+}
+
+fn cell_to_string(row: &Row, column: &Column, idx: usize) -> String {
```
Comment (melgenek) [inline datafusion/core/tests/sqllogictests/src/engines/postgres/mod.rs:304]: There are no changes to this function, I just moved it downwards.

File: datafusion/core/tests/sqllogictests/src/engines/postgres/mod.rs [hunk 4]
```diff
@@ -290,27 +249,20 @@ impl sqllogictest::AsyncDB for Postgres {
...
-            .collect::<Vec<_>>();
 
-        if output.is_empty() {
-            let stmt = self.client.prepare(sql).await?;
-            Ok(DBOutput::Rows {
-                types: vec![DFColumnType::Any; stmt.columns().len()],
-                rows: vec![],
-            })
+        if rows.is_empty() {
+            return Ok(DBOutput::StatementComplete(0));
```
Comment (melgenek) [inline datafusion/core/tests/sqllogictests/src/engines/postgres/mod.rs:254]: Datafusion currently considers empty results as "statements" https://github.com/apache/arrow-datafusion/blob/master/datafusion/core/tests/sqllogictests/src/engines/datafusion/normalize.rs#L32-L35.
This change makes Postgres produce the same result when there are no values.

Technically, an opposite change is possible: update Datafusion to consider an empty batch as empty query results, but with types.

File: datafusion/core/tests/sqllogictests/test_files/decimal.slt [hunk 5]
```diff
@@ -524,7 +524,7 @@ select * from decimal_simple where c1 >= 0.00004 order by c1 limit 5;
 0.00005 0.000000000005 9 true 0.000052
 
 
-query RRI?R
+query FFIBF
```
Comment (melgenek) [inline datafusion/core/tests/sqllogictests/test_files/decimal.slt:527]: I updated types using `--complete` in this pr. I pointed datafusion to a local version of `sqllogictest-rs` with changes from this pr https://github.com/risinglightdb/sqllogictest-rs/pull/164. `sqllogictest-rs:0.12` doesn't complete types.

File: datafusion/core/tests/sqllogictests/test_files/pg_compat/pg_compat_type_coercion.slt [hunk 6]
```diff
@@ -15,108 +15,37 @@
...
 # under the License.
 
-query ??
+query BB
 select true and true, false and false;
 ----
 true false
 
 
-onlyif DataFusion
```
Comment (melgenek) [inline datafusion/core/tests/sqllogictests/test_files/pg_compat/pg_compat_type_coercion.slt:24]: The `arrow_typeof` and `pg_typeof` checks are no longer needed for the coercion tests. Sqllogictest now checks that the columns are `Boolean`.

File: datafusion/core/tests/sqllogictests/src/engines/postgres/mod.rs (hunk 4 above)
Comment (alamb) [reply datafusion/core/tests/sqllogictests/src/engines/postgres/mod.rs:254]: The opposite (empty result as a query result with schema) is probably a better behavior as it would allow checking types for statements like `COUNT(*) FROM foo WHERE false`

**Review by alamb — COMMENTED** [review]
Thank you so much @melgenek  this is looking great. This feature is so cool!

Prior to merging this PR, I would like to suggest:
1. Update the type characters to be more inline with other systems
2. Wait for a release of `sqllogictests-rs` that includes the complete feature (as I think it is critical to the maintenance of the sqllogictests in datafusion)

cc @xudong963

File: datafusion/core/tests/sqllogictests/README.md (hunk 1 above)
Comment (alamb) [reply datafusion/core/tests/sqllogictests/README.md:114]: I think we should stick as close as possibly to other standards -- so therefore I suggest:

 `T` for text
 `P` for timestamP

I also think it would be good to differentiate between `F` and `R` if possible 

What do you think?

**Comment by xudong963** [comment]
Cool, I'm traveling. Will review it as soon as possible.

File: datafusion/core/tests/sqllogictests/src/engines/postgres/mod.rs (hunk 4 above)
Comment (melgenek) [reply datafusion/core/tests/sqllogictests/src/engines/postgres/mod.rs]: Makes sense. I updated the implementation here https://github.com/apache/arrow-datafusion/pull/5253/commits/e9733d76179f2a9cbb841d94f6239c12ab7115b8.

As a side note, I noticed that `sqllogictest-rs` doesn't implement checks when the expected record is a statement, for example, `statement ok`. In this case, even if the client implementation returns results, neither types nor actual rows would be checked.

File: datafusion/core/tests/sqllogictests/README.md (hunk 1 above)
Comment (melgenek) [reply datafusion/core/tests/sqllogictests/README.md]: > I think we should stick as close as possibly to other standards -- so therefore I suggest:
> T for text
> P for timestamP

Sounds good. It would probably make it easier to use tests suites from other databases, if you ever want to run them.
Updated to be https://github.com/apache/arrow-datafusion/pull/5253/commits/7e32de81b1c238cd5932de707f12f8f40a0c9c9b.

> I also think it would be good to differentiate between F and R if possible

This one is trickier. It would be hard to check Postgres/Datafusion compatibility on floating points then. Postgres usually falls back to using `Numeric` by default, which is their decimal. Whereas Datafusion prefers Floats unless otherwise specified.

I could technically provide a custom validator for postgres that treats expected results the same way for Postgres. So the Postgres' validator would treat `query FR` the same way as `query RR`. 
If you don't think having different validators is confusing, I can do it.

**Comment by melgenek** [comment]
> Wait for a release of sqllogictests-rs that includes the complete feature (as I think it is critical to the maintenance of the sqllogictests in datafusion)

👍

File: datafusion/core/tests/sqllogictests/README.md (hunk 1 above)
Comment (alamb) [reply datafusion/core/tests/sqllogictests/README.md]: > This one is trickier. It

If it is tricky I don't think it is needed then

**Review by alamb — APPROVED** [review]
LGTM -- thanks @melgenek  -- no we just need to wait on the sqllogictest release / upgrade and we'll be good to go

**Comment by alamb** [comment]
Marking as draft until the upstream release is ready

**Comment by melgenek** [comment]
@alamb I updated the `sqllogictest-rs` to `0.13.0`. The `--complete` option fills in the types now. 
In fact, you can now write `query\n` without specifying any types at all and completion would work.

This pr is good to go.

**Review by alamb — APPROVED** [review]
Thank you so much @melgenek  

I merged this PR up to the latest main and ran `cargo test --test sqllogictests  -- --complete`  and there are no changes 👍 

Then I also tried purposely updating test output to be incorrect, and I got an error as expected


```
Error: query columns mismatch:
[SQL] SELECT * from information_schema.tables;
[Expected] [R][R]
[Actual  ] [T][T][T][T]
at tests/sqllogictests/test_files/information_schema.slt:33
```

And then `cargo test --test sqllogictests  -- --complete` fixed the problem I introduced

❤️ 🚀  -- thank you
